Work Queue Slack App tutorial (#259)
temporalio/temporal-learning: docs/tutorials/typescript/digital-ocean-deployment/index.mdx

---
id: index
title: Deploy a TypeScript application to production on a Digital Ocean Droplet
description: Learn how to deploy your TypeScript-based Temporal Application to a DigitalOcean Droplet using Ubuntu 20.04 or 22.04
sidebar_label: Deploy a TypeScript application to production on a Digital Ocean Droplet
tags:
  - typescript
  - deployment
  - digital-ocean
  - temporal-cloud
keywords:
  - TypeScript Temporal Application deployment
  - DigitalOcean Droplet setup
  - Ubuntu 20.04 server setup
  - Ubuntu 22.04 server setup
  - Install Node.js on Ubuntu
  - Install TypeScript on Ubuntu
  - SSH setup DigitalOcean
  - Custom domain configuration
  - Nginx reverse proxy setup
  - Let's Encrypt SSL certificate
  - PM2 process manager
  - Temporal Cloud integration
  - Deploy TypeScript application DigitalOcean
  - Temporal Cloud certificates
---

When you are ready to deploy your TypeScript based Temporal Application to production, you need a server to host it.

> Digital Ocean is provides flexible Cloud servers called Droplets that can be used to host your application.

Reviewer: ```suggestion
Digital Ocean provides flexible Cloud servers called Droplets that can be used to host your application.
```

Typo here with "is provides," but also -- I'm not sure we need to single out DigitalOcean here as a hosting provider. Why don't we just turn this tutorial into "How to Deploy a Temporal Application to Production"?

Author: For the same reason we would call out Slack or any other technology in the title of the tutorial, SEO. Tailoring this specifically to Digital Ocean is a major objective here.

> In this tutorial we will walk through the steps to deploy your Temporal Application to a Digital Ocean Droplet while using Temporal Cloud as your Temporal Application's supervisor.

Reviewer: "supervisor" isn't language I've seen often here, and if this is a Cloud tutorial, I think that should be in the title, not DigitalOcean.

Moreover -- "Ensure you can connect to the Temporal Service" should probably be a prerequisite tutorial that we have on hand and we can link to before we publish this tutorial. Perhaps one option for self-hosted and one option for Cloud? That way this is scoped better.

Author: I disagree on the first point and to be honest I don't understand the second point.

Reviewer: OK, understood about the SEO argument for the first point.

For the second point: I think there's a lot of general "make sure you can SSH / make sure the Cloud handshake works in the most basic ways" stuff in this tutorial. If you'd rather keep it in, I understand why, but then I think we should reframe the introduction here slightly so that the audience knows that this is all about being introduced to Cloud.

## Create a Droplet {#create-droplet}

> Assuming you already have a Digital Ocean Account, create a new Droplet and choose the Ubuntu 20.04 image or Ubuntu 22.04 image.

Reviewer: "Assuming you already have a DigitalOcean account" is not good scoping for us -- way too much is elided in this one sentence (see above comments).

Author: Sounds like having a Digital Ocean account should be a prereq.

**Set up SSH**

Next, set up SSH so you can set up your server from your local terminal.
You can do that by following the [Add SSH Key to Droplet tutorial](https://docs.digitalocean.com/products/droplets/how-to/add-ssh-keys/).

If you already have a private key on your local machine and you need to force SSH to use the new key, you can force it with the `-i` flag, for example:

```
ssh -i ~/.ssh/id_rsa_digitalocean
```

> **Install Node.js**

Reviewer: OK, here's where I think the meat of this one should begin after prerequisites and after some re-scoping. I think it's OK for us to link to DigitalOcean tutorials here a little bit, but I'd still encourage us to provide a bit more context here:

Since you are going to be deploying a TypeScript application, you need to install TypeScript into this environment. First, install Node.js, a JavaScript runtime for server-side programming.

There are several options for installing Node.js. The most straightforward is to install from your system package manager. For example, on Ubuntu:

```
sudo apt update
sudo apt install nodejs
```

This will install the newest stable version from Ubuntu sources. If you need to install a different version of Node.js for compatibility reasons, refer to [this DigitalOcean tutorial](https://www.digitalocean.com/community/tutorials/how-to-install-node-js-on-ubuntu-22-04).

Author: I still think the tutorial should include ssh steps.
Managing the SSH connection was a constant and critical thing.
I feel like the whole experience depends on maintaining that connection.

Once you have SSH'd into your Droplet, you can install Node.js.

Follow these tutorials, for Ubuntu 20.04 or 22.04 respectively:

- [How To Install Node.js on Ubuntu 20.04](https://www.digitalocean.com/community/tutorials/how-to-install-node-js-on-ubuntu-20-04)
- [How To Install Node.js on Ubuntu 22.04](https://www.digitalocean.com/community/tutorials/how-to-install-node-js-on-ubuntu-22-04)

**Install TypeScript and TS Node**

> 

Reviewer: ```suggestion
Next, use the Node.js package manager, `npm`, to install some packages that provide TypeScript compatibility. Install the `typescript` and `ts-node` packages globally, using the `-g` flag:

```

```
sudo npm install -g typescript ts-node
```

## Use a custom domain {#custom-domain}

> You need a domain, because you need to have a valid SSL certificate for the Temporal Application to communicate with Temporal Cloud.

Reviewer: This should also be in its own bullet higher up, under prerequisites -- it's not part of the main configuration and it's out of scope to get into actually configuring LetsEncrypt here, but we should say something like:

If you are deploying into a production environment, this server should have a domain name assigned so it can be configured with an SSL certificate. If you have already assigned a domain to this server, you can refer to the [LetsEncrypt documentation](https://certbot.eff.org/instructions) to configure a free SSL certificate.

Author: I moved the domain up to prereqs... but pretty sure you can't issue a cert until the nginx server block is set up - so that step is blocked by one in the tutorial.

Reviewer: yeah, OK! I think I'd still like to see us refer to the LetsEncrypt docs here in a more general so we're not leaning on DO content quite so heavily, but that makes sense.

Register your domain and ensure the domain's A record points to the droplet's IP address.

> ## Set up Nginx as a reverse proxy {#set-up-nginx}

Reviewer: ```suggestion
## Step 2 -- Install and Configure Nginx as a Reverse Proxy {#set-up-nginx}
```

After looking at how our h2s are displayed in Docusaurus, I think we absolutely need some kind of numbering -- otherwise it makes it much too hard (imo) to read these pages linearly, they wind up looking like reference documentation. I think this would technically be Step 2, if Step 1 is installing Nodejs and everything else is prerequisite.

> SSH into your Droplet and set up Nginx as a reverse proxy.

Reviewer: I don't think we need to keep reminding people to SSH through this tutorial -- let's try to keep in mind what level of complexity this is targeted at. We assume that people know enough at this point to connect to their production environment and need to be walked through the actual connection steps that pertain to this deployment.

```suggestion
Because you'll be deploying an application that handles external requests over HTTP/S, we strongly recommend also deploying an Nginx reverse proxy. 

Node.js applications typically bind to `localhost`. This means by default, your application will only be accessible locally on the machine it resides on. While you can specify a different port to allow access through the internet, Node.js's built in application server provides fairly minimal security and configurability. Using a reverse proxy isolates the application server from direct internet access, allows you to centralize firewall protection, and minimizes the attack surface for common threats such as denial of service attacks.

From a client’s perspective, interacting with a reverse proxy is no different from interacting with the application server directly. It is functionally the same, and the client cannot tell the difference. A client requests a resource and then receives it, without any extra configuration required by the client. Nginx, a popular open source web server, can act as a reverse proxy with a nearly drop-in configuration.
```

Follow these tutorials, for Ubuntu 20.04 or 22.04 respectively, including setting up the server blocks:

- [Set up Nginx for Ubuntu 20.04](https://www.digitalocean.com/community/tutorials/how-to-install-nginx-on-ubuntu-20-04)
- [Set up Nginx for Ubuntu 22.04](https://www.digitalocean.com/community/tutorials/how-to-configure-nginx-as-a-reverse-proxy-on-ubuntu-22-04)

> ## Create a domain certificate {#create-domain-cert}

Reviewer: Remove -- handled above (and should be moved to prereqs)

Author: The problem is that you can't do this step until you have setup a server block with a ngnix config.

So you can't move domain cert to prereqs - because it depends on this step.

Use Let's encrypt certbot to create a certificate for your domain.
This will effectively enable HTTPS for your domain and your Digital Ocean Droplet.

Follow the tutorial for your Ubuntu version:

- [How To Secure Nginx with Let's Encrypt on Ubuntu 20.04](https://www.digitalocean.com/community/tutorials/how-to-secure-nginx-with-let-s-encrypt-on-ubuntu-20-04)
- [How To Secure Nginx with Let's Encrypt on Ubuntu 22.04](https://www.digitalocean.com/community/tutorials/how-to-secure-nginx-with-let-s-encrypt-on-ubuntu-22-04)

> ## Create a Temporal Cloud certificate {#create-temporal-cloud-cert}

Reviewer: ```suggestion
## Step 3 -- Create a Temporal Cloud certificate {#create-temporal-cloud-cert}
```

This section is pretty good, well-motivated, concise, and links out to other content of ours. All good things.

You will need a seperate certificate to authenticate the communication between your Temporal Client, Temporal Workers, and Temporal Cloud.

You can use tcld to generate the certificate.

Follow the [steps to install tcld](https://docs.temporal.io/cloud/tcld#install-tcld).

Then, follow the [steps to generate a Temporal Cloud certificate using tcld](https://docs.temporal.io/cloud/certificates#use-tcld-to-generate-certificates).

> ## Clone your repo into the droplet {#clone-repo}

Reviewer: ```suggestion
## Step 4 -- Deploy your Application to the Server {#clone-repo}
```

Git just became an additional dependency here, assuming people are deploying via Git -- I think we can provide a little bit of boilerplate here around `git clone` and `npm i` since those are pretty standard for TS applications.

Clone the repo on the droplet

Install dependencies

> ## Create a `.env` file

Reviewer: I think this should probably be a subsection (h3?) of Step 4, but I don't know how rigidly we're enforcing the use of these environment variables in TS clients, so I think we need a bit more context here for how these are being used.

Add your Temporal Cloud Namespace, address, certificate PEM and KEY to your projects `.env` file.

```plaintext
TEMPORAL_CLOUD_ADDRESS=""
TEMPORAL_CLOUD_NAMESPACE=""
TEMPORAL_CLOUD_PEM=""
TEMPORAL_CLOUD_PRIVATE_KEY=""
```

Your Temporal Client and Worker will need those environment variables to connect to Temporal Cloud.

> ## Ensure you are using Temporal Cloud {#use-temporal-cloud}

Reviewer: Another scoping comment -- all of a sudden here we're talking about migrating an application to use Temporal Cloud. I'm not sure whether this belongs in a separate reference article (we could easily publish one for each language), linked under Step 4 above, or as a Step 5 here, but I think we need to more consistent in how we're framing this.

Author: It is not migrating though.
You can't use anything but Temporal Cloud when deploying to a DO server like this - 
Ensuring there are proper env vars and they and the code is structured to use them is a critical step - I don't think farming it out to another tutorial is the way to go here -

Author: I do get where you are coming from here tho - it feels like it could be a prereq - 
For the scope I am pushing for, I want to keep it for now.

The current tutorial for running Workers with TypeScript feels heavy https://learn.temporal.io/getting_started/typescript/run_workers_with_cloud_typescript/

Reviewer: I mean, technically you *could* run a self-hosted Temporal on the same droplet as your workers, right? That's what our upcoming self-hosted tutorial is going to look like. Just want to make sure we're drawing that distinction effectively.

```ts
import "dotenv/config";
import path from "path";
import {Worker, NativeConnection} from "@temporalio/worker";
import * as activities from "./activities/index";

async function run() {
  try {
    const key = Buffer.from(
      process.env.TEMPORAL_CLOUD_PRIVATE_KEY || "",
      "utf-8"
    );
    const cert = Buffer.from(process.env.TEMPORAL_CLOUD_PEM || "", "utf-8");
    const connection = await NativeConnection.connect({
      address: process.env.TEMPORAL_CLOUD_ADDRESS || "",
      tls: {
        clientCertPair: {
          crt: cert,
          key: key,
        },
      },
    });

    const worker = await Worker.create({
      connection,
      namespace: process.env.TEMPORAL_CLOUD_NAMESPACE || "",
      workflowsPath: path.resolve(__dirname, "./workflows"),
      activities,
      taskQueue: `${process.env.ENV}-temporal-iq-task-queue`,
    });

    await worker.run();
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
}

run();
```

<!--SNIPEND-->

> ## Use `pm2` to run your Worker {#use-pm2}

Reviewer: ```suggestion
## Step 5 -- Use `pm2` to run your Worker {#use-pm2}
```

Need one more sentence here about what this does, eg, pm2 is a popular way to daemonize JavaScript and TypeScript applications to run in perpetuity.

```shell
sudo npm install -g pm2
```

Change directory into your project and run your Temporal Worker with `pm2`.

```shell
pm2 start <your-app-entry>.ts --interpreter ts-node
```

## Conclusion

> You should now have a TypeScript based Temporal Application running on a Digital Ocean Droplet using Temporal Cloud as your Temporal Application's supervisor.

Reviewer: See earlier notes about whether we need this to be DO-specific at all -- but I think we have a solid 5 steps and then a conclusion, after we've moved some stuff into prereqs.
